apps/src/blockly/addons/cdoFieldPicker.js

```diff
@@ -0,0 +1,129 @@
+import GoogleBlockly from 'blockly/core';
+
+const BUTTON_CORNER_RADIUS = 3;
+const BUTTON_INNER_HEIGHT = 16;
+import {DEFAULT_SOUND} from '@cdo/apps/blockly/constants';
+
+/**
+ * This is a customized field which the user clicks to select an option from a customized picker,
+ * for example, the location of a sprite from a grid or a sound file from a customized modal.
+ * @param value The initial value of the field.
+ * @param onChange The function that handles the field's editor.
+ * @param onDisplay The function tht handles how the field text is displayed.
+ * @param buttonIcon If the field displyas a button, this is the string that is displayed on the button.
+ */
+export default class CdoFieldPicker extends GoogleBlockly.Field {
+  constructor(value, onChange, onDisplay, buttonIcon) {
+    super(value);
+    this.onChange = onChange;
+    this.onDisplay = onDisplay;
+    this.buttonIcon = buttonIcon;
+
+    this.SERIALIZABLE = true;
+  }
+
+  static fromJson(options) {
+    return new CdoFieldPicker(
+      options.value,
+      options.onChange,
+      options.onDisplay,
+      options.buttonIcon
+    );
+  }
+
+  /**
+   * Validate the changes to a field's value before they are set.
+   * @override
+   */
+  doClassValidation_(newValue) {
+    if (typeof newValue !== 'string') {
+      return null;
+    }
+    // Handle 'play sound' block with default param from CDO blockly.
+    // TODO: Remove when sprite lab is migrated to Google blockly.
+    if (newValue === 'Choose') {
+      return DEFAULT_SOUND;
+    }
+    return newValue;
+  }
+
+  /**
+   * Create the block UI for this field.
+   * @override
+   */
+  initView() {
+    super.initView();
+    if (this.buttonIcon) {
+      this.buttonElement_ = Blockly.utils.dom.createSvgElement(
+        'rect',
+        {
+          rx: BUTTON_CORNER_RADIUS,
+          ry: BUTTON_CORNER_RADIUS,
+          x: 1,
+          y: 1,
+          height: BUTTON_INNER_HEIGHT,
+          width: BUTTON_INNER_HEIGHT,
+        },
+        this.fieldGroup_
+      );
+      this.buttonElement_.style.fillOpacity = 1;
+      this.buttonElement_.style.fill =
+        this.getSourceBlock().style.colourSecondary;
+
+      this.textElement_.style.fontSize = '11pt';
+      this.textElement_.style.fill = this.getSourceBlock().style.colourPrimary;
+      this.textElement_.textContent = '';
+      this.textElement_.appendChild(this.buttonIcon);
+
+      this.fieldGroup_.insertBefore(this.buttonElement_, this.textElement_);
+    }
+  }
+
+  /**
+   *  Get the text from this field to display on the block. May differ from
+   * `getText` with call to `this.onDisplay` which can change format of text.
+   * @override
+   */
+  getDisplayText_() {
```

Comment:
Is this for the block text that shows location coordinates or for the field text that shows the sound name?

Reply:
It depends on how `this.onDisplay` is defined as passed from the constructor parameter.
For the sound picker, it is the field text. For the location picker, it is the label text next to the field.